backend/FwLite/LcmCrdt/CrdtMiniLcmApi.cs

```diff
@@ -237,21 +282,97 @@ await dataModel.AddChanges(ClientId,
+                {
+                    throw new InvalidOperationException($"Complex form component {complexFormComponent} references deleted entry {complexFormComponent.ComplexFormEntryId} as its complex form");
+                }
+
+                if (complexFormComponent.ComponentSenseId != null &&
+                    !await Senses.AnyAsyncEF(s => s.Id == complexFormComponent.ComponentSenseId.Value))
+                {
+                    throw new InvalidOperationException($"Complex form component {complexFormComponent} references deleted sense {complexFormComponent.ComponentSenseId} as its component");
+                }
+                yield return new AddEntryComponentChange(complexFormComponent);
+            }
+        }
+
+        async IAsyncEnumerable<AddComplexFormTypeChange> ToComplexFormTypes(IList<ComplexFormType> complexFormTypes)
+        {
+            foreach (var complexFormType in complexFormTypes)
+            {
+                if (complexFormType.Id == default)
+                {
+                    throw new InvalidOperationException("Complex form type must have an id");
+                }
+
+                if (!await ComplexFormTypes.AnyAsyncEF(t => t.Id == complexFormType.Id))
+                {
+                    throw new InvalidOperationException($"Complex form type {complexFormType} does not exist");
+                }
+                yield return new AddComplexFormTypeChange(entry.Id, complexFormType);
+            }
+        }
     }
 
+    private async ValueTask<bool> IsEntryDeleted(Guid id)
+    {
+        return !await Entries.AnyAsyncEF(e => e.Id == id);
+    }
+
+
+
     public async Task<Entry> UpdateEntry(Guid id,
         UpdateObjectInput<Entry> update)
     {
         var entry = await GetEntry(id);
         if (entry is null) throw new NullReferenceException($"unable to find entry with id {id}");
 
         await dataModel.AddChanges(ClientId, [..entry.ToChanges(update.Patch)]);
-        return await GetEntry(id) ?? throw new NullReferenceException();
+        return await GetEntry(id) ?? throw new NullReferenceException("unable to find entry with id " + id);
+    }
+
+    public async Task<Entry> UpdateEntry(Entry entry)
+    {
+        var commitId = entry.GetVersionGuid();
+
+        var commitHybridDate = await dbContext.Set<Commit>()
+            .Where(c => c.Id == commitId)
+            .Select(c => c.HybridDateTime)
+            .SingleAsyncEF();
+        //todo this will not work for properties not in the snapshot, but we don't have a way to get the full snapshot yet
+        //todo also, add api for getting an entity at a specific commit
+        var snapshot = await dataModel.GetEntitySnapshotAtTime(commitHybridDate.DateTime.AddSeconds(1), entry.Id);
+        var before = snapshot?.Entity.DbObject as Entry;
+        ArgumentNullException.ThrowIfNull(before);
+        //workaround to avoid syncing senses, which are not in the snapshot
```

Comment:
I'm not sure how to resolve this issue

Reply:
So we basically have to replicate this?
![image](https://github.com/user-attachments/assets/a2675430-4325-4384-bc0a-73bc5b325d55)

Reply:
sort of, except that is querying the projected tables, which only contain the current state. We would basically need to pull the entire project into memory by building up commits and then find everything that references our entry by ID